src/shogun/lib/SGMatrix.h

```diff
@@ -527,5 +553,264 @@ template<class T> class SGMatrix : public SGReferencedData
+		{
+			return m_col_idx > other.m_col_idx;
+		}
+		bool operator<=(const column_iterator& other) const
+		{
+			return m_col_idx <= other.m_col_idx;
+		}
+		bool operator>=(const column_iterator& other) const
+		{
+			return m_col_idx >= other.m_col_idx;
+		}
+
+		difference_type operator-(const column_iterator& other) const
+		{
+			return m_col_idx - other.m_col_idx;
+		}
+
+	private:
+		SGMatrix<T>& m_mat;
+		difference_type m_col_idx;
+	};
+	template<typename T>
+	class SGMatrix<T>::const_column_iterator
+	{
+	public:
+		// Iterator traits
+		using difference_type = index_t;
+		using value_type = const SGVector<T>;
+		using pointer = const SGVector<T>*;
+		using reference = const SGVector<T>&;
+		using iterator_category = std::random_access_iterator_tag;
+
+		const_column_iterator(const SGMatrix<T>& mat, index_t idx = 0)
+			: m_mat(mat), m_col_idx(idx)
+		{
+		}
+		const_column_iterator(const const_column_iterator& other)
+			: m_mat(other.m_mat), m_col_idx(other.m_col_idx)
+		{
+		}
+
+		const_column_iterator& operator++()
+		{
+			++m_col_idx;
+			return *this;
+		}
+		const_column_iterator operator++(int)
+		{
+			const_column_iterator retval(*this);
+			++(*this);
+			return retval;
+		}
+		const_column_iterator& operator--()
+		{
+			--m_col_idx;
+			return *this;
+		}
+		const_column_iterator operator--(int)
+		{
+			const_column_iterator retval(*this);
+			--(*this);
+			return retval;
+		}
+
+		bool operator==(const const_column_iterator& other) const
+		{
+			return m_mat == other.m_mat && m_col_idx == other.m_col_idx;
+		}
+		bool operator!=(const const_column_iterator& other) const
+		{
+			return !(*this == other);
+		}
+
+		value_type operator*()
+		{
+			return m_mat.get_column(m_col_idx);
+		}
+
+		const_column_iterator& operator+=(difference_type d)
+		{
+			m_col_idx += d;
+			return *this;
+		}
+		const_column_iterator& operator-=(difference_type d)
+		{
+			m_col_idx -= d;
+			return *this;
+		}
+
+		const_column_iterator operator+(difference_type d) const
+		{
+			return const_column_iterator(m_mat, m_col_idx + d);
+		}
+		const_column_iterator operator-(difference_type d) const
+		{
+			return const_column_iterator(m_mat, m_col_idx - d);
+		}
+
+		value_type operator[](difference_type d) const
+		{
+			return *const_column_iterator(m_mat, m_col_idx + d);
+		}
+
+		bool operator<(const const_column_iterator& other) const
+		{
+			return m_col_idx < other.m_col_idx;
+		}
+		bool operator>(const const_column_iterator& other) const
+		{
+			return m_col_idx > other.m_col_idx;
+		}
+		bool operator<=(const const_column_iterator& other) const
+		{
+			return m_col_idx <= other.m_col_idx;
+		}
+		bool operator>=(const const_column_iterator& other) const
+		{
+			return m_col_idx >= other.m_col_idx;
+		}
+
+		difference_type operator-(const const_column_iterator& other) const
+		{
+			return m_col_idx - other.m_col_idx;
+		}
+
+	private:
+		const SGMatrix<T>& m_mat;
+		difference_type m_col_idx;
+	};
+#ifndef SWIG
```

Comment:
note that all of the iterator code should be hidden from swig, not just parts

Reply:
Ah cool, btw should I write "SWIG Should skip this part". I mean it's apparent anyway, thoughts?

Reply:
no need imo